src/Swift.Runtime/src/Metadata/TypeMetadata.cs

```diff
@@ -249,4 +253,125 @@ public override int GetHashCode ()
     /// Gets the type metadata cache for the runtime.
     /// </summary>
     public static ITypeMetadataCache Cache => cache;
-}
\ No newline at end of file
+
+    /// <summary>
+    /// Attempt to get the Swift type metadata for the given object instance
+    /// </summary>
+    /// <typeparam name="T">The type of the object</typeparam>
+    /// <param name="obj">An object to look up</param>
+    /// <param name="result">The result of looked up type metadata</param>
+    /// <returns>true on success false otherwise</returns>
+    public static bool TryGetTypeMetadata<T>(T obj, [NotNullWhen(true)] out TypeMetadata? result)
+    {
+        return TryGetTypeMetadata(typeof(T), out result);
+    }
+
+    /// <summary>
+    /// Attemp to get the Swift type metadata for the given type
+    /// </summary>
+    /// <param name="type">A type to look up</param>
+    /// <param name="result">The result of looked up type metadata</param>
+    /// <returns>true on success false otherwise</returns>
+    public static bool TryGetTypeMetadata(Type type, [NotNullWhen(true)] out TypeMetadata? result)
+    {
+        if (cache.TryGet(type, out result))
+            return true;
+
+        return TryGetTypeMetadataUncached(type, out result);
+    }
+
+    /// <summary>
+    /// Attempt to get the Swift type metadata for the given type
+    /// </summary>
+    /// <param name="type">A type to look up</param>
+    /// <returns>The result of the looked up type metadata on success</returns>
+    /// <exception cref="Exception">Throws when lookup fails</exception>
+    public static TypeMetadata GetTypeMetadataOrThrow(Type type)
+    {
+        if (TryGetTypeMetadata(type, out var result))
+            return result.Value;
+        throw new Exception(string.Format("Unable to get type metadata for type {0}", type.Name));
+    }
+
+    /// <summary>
+    /// Attempt to get the Swift type metadata but without accessing the cache
+    /// </summary>
+    /// <param name="type">A type to look up</param>
+    /// <param name="result">The result of looked up type metadata</param>
+    /// <returns>true on success false otherwise</returns>
+    /// <exception cref="NotImplementedException">Throws if unable to look up the ISwiftObject.GetTypeMetadata method.</exception>
+    static bool TryGetTypeMetadataUncached(Type type, [NotNullWhen(true)] out TypeMetadata? result)
```

Comment:
I think this is not trimming / NativeAOT friendly and brings us back to the discussion we had in https://github.com/dotnet/runtimelab/pull/2771. I would recommend going with the pattern described in binding-generics. 

cc: @MichalStrehovsky, @vitek-karas

Reply:
We should add `<IsAotCompatible>true</IsAotCompatible>` to the project file (that will also enable IsTrimmable, etc.) so that problematic code gets surfaced as warnings.

Reply:
I need some help here - what do you want me to do with this?
I can remove the reflection code and make a static initializer for the type that forces it into the cache.

Reply:
Well the simplest "fix" would be to make this method generic - that means that `TryGetTypeMetadata` will need to be generic as well and use typeof(T) to populate cache.

Then we would need something like
```csharp
public static class Helper<T> where T : ISwiftObject 
{
    public static TypeMetadata GetTypeMetadata()
    {
        return T. GetTypeMetadata();
    }
}
```

and in the body of the method you'd rewrite the reflection call to:
```csharp
if (typeof(ISwiftObject).IsAssignableFrom(typeof(T)))
{
    var helperType = typeof(Helper<>).MakeGenericType(typeof(T));
    result = (TypeMetadata)helperType.GetMethod("GetTypeMetadata")!.Invoke(null, null)!;
    return true;
}
```

This would make the trimming errors disappear.

I agree that this is not the prettiest solution, it is probably quite slow as well - but we dont have anything better now. Maybe in the future we could push the csharp lang team to give as `type is someOtherType` pattern in c#. 

Alternative would be to make the MetadataAccessors non-static and use `x is ISwiftObject` pattern. But this will cause a lot of weirdness when we need to get metadata in order to create objects instance.

Reply:
Yeah, unfortunately, one of the most common cases of getting a type's metadata is when we have no instance.